Use a dictionary for Postgres row column lookups.

MrMage opened the pull request:
It appears that row decoding with `NIOPostgres` on Swift 5 is already way faster than `Postgres` on Swift 4. I suspect this is partly due to Swift 5's small strings making column lookups and string comparisons _way_ faster than before (reducing the impact of a linear search vs a hash table lookup), and partly due to `NIOPostgres` no longer having to rebuild its column lookup table for every single row.

~**However, I noticed that my Postgres database returns rows in `text` encoding rather than `binary`.** @tanner0101, this is something you might want to look into.~

Results of this PR:
- Performance for small, medium and large models is nearly identical
- `testSelectLargeModelWithLongFieldNames` runs ~26% faster
- `testSelectHugeModel` funs ~53% faster

tanner0101 commented:
@MrMage use `query` instead of `simpleQuery` to select rows using binary encoding. You can use `simpleQuery` to do all of the table management and inserts, but use `query` for the SELECTs at least.

tanner0101 commented:
>However, I noticed that my Postgres database returns rows in text encoding rather than binary. @tanner0101, this is something you might want to look into. 

@MrMage use `query` instead of `simpleQuery` to select rows using binary encoding. You can use `simpleQuery` to do all of the table management and inserts, but use `query` for the SELECTs at least.

tanner0101 commented on Sources/NIOPostgres/Data/PostgresRow.swift:
```diff
@@ -2,50 +2,64 @@ public struct PostgresRow: CustomStringConvertible {
+            for (fieldIndex, field) in rowDescription.fields.enumerated() {
+                columnNameToIndexLookupTable[field.name, default: []].append(.init(indexInRow: fieldIndex, field: field))
+            }
+            self.columnNameToIndexLookupTable = columnNameToIndexLookupTable
```
Should we populate this lookup table lazily? We could make the lookup table optional, then populate it on the first call to `lookup` when it is `nil`. That way we don't add any extra computation where the results just end up being discarded anyway.

MrMage replied:
Done.

tanner0101 commented on Sources/NIOPostgres/Data/PostgresRow.swift:
```diff
@@ -2,50 +2,64 @@ public struct PostgresRow: CustomStringConvertible {
             self.rowDescription = rowDescription
             self.resultFormat = resultFormat
+
+            var columnNameToIndexLookupTable: [String: [Entry]] = [:]
```
I don't think we need to store this as an array since `lookup` just calls `first` anyway. If we ever add support for getting all values for a given column name we could change this. But IIRC Fluent 3 only lets you get the first column anyway, and I think Fluent 4 should follow suit.

MrMage replied:
if `tableOID != 0`, we actually check a different column name:

```
            if tableOID == 0 {
                return columnTable.first
            } else {
                return columnTable.first { $0.field.tableOID == tableOID }
            }
```

I think we should keep this; otherwise fetching joined tables with two identical column names will likely fail.

tanner0101 replied:
Ah okay, missed that part. I think we should consider removing support for getting a column for a specific tableOID. That can be a separate PR though so I don't slow things down here. Thanks!

tanner0101 commented on Sources/NIOPostgres/Data/PostgresRow.swift:
```diff
@@ -2,50 +2,64 @@ public struct PostgresRow: CustomStringConvertible {
             rowDescription: PostgresMessage.RowDescription,
             resultFormat: [PostgresFormatCode]
-        ) {
+            ) {
```
spacing

MrMage commented:
I have updated the PR; please take another look.

> Regarding the test failure, maybe we should remove `testRangeSelectPerformance`? Or potentially we could increase the amount of times the test runs inside the measure block to reduce std deviation?

I have removed two fairly meaningless tests (they benchmark the Postgres server more than Vapor) and increased the iteration count on the third test now.

MrMage commented:
Seems like the CI failure is unrelated to this PR.

tanner0101 commented:
> Seems like the CI failure is unrelated to this PR.

Strange... running the tests again seems to have fixed it. I'll keep my eye out for that one.

penny-coin commented:
Hey @MrMage, you just merged a pull request, have a coin! 

You now have 28 coins.
